### DISCUSSION TITLE
Genes returned by getMarkers() vs markerHeatmap()

### DISCUSSION THREAD

**Discussion by oligomyeggo** [GreenleafLab/ArchR#1503] [Questions / Documentation]
Hello, and thank you for an incredible package to analyze scATAC-seq data! I have been following your detailed manual with my own data and it has been working great. I have a question about some of the data generated from [section 7.3,, Identifying Marker Genes](https://www.archrproject.com/bookdown/identifying-marker-genes.html), namely the output from running `getMarkers()` and `markerHeatmap()`. Both functions take in the `SummarizedExperiment` object generated from running the `getMarkerFeatures()` function, and for both functions I am using the same `cutOff` values. 

From my understanding, the `DataFrame` objects returned from running `getMarkers()` returns a `DataFrame` for each cluster with marker genes sorted by `FDR`. I am struggling to understand how the top `n` genes are selected per cluster by the `markerHeatmap()` function, though. They do not match the order of the genes in the data frames (sorted by `FDR`), and they also don't seem to match the order of the genes in the data frames if I sort by other metrics (such as `MeanDiff` or `Log2FC`). Also, for one of my clusters, I am only getting back one gene that matches the `cutoff` values after running `getMarkers()`. However, I am getting `n=15` genes back for this cluster when running `markerHeatmap()`. 

I see in the code that both `getMarkers()` and `markerHeatmap()` start off by generating the same `passMat` object, but that the `passMat` is manipulated a little more in the `markerHeatmap()` function and is used to generated a `spmat` object which is ultimately used to get the heatmap marker genes. I think I am just missing something in the `markerHeatmap()` code, but any insight into how the top marker genes for the `markerHeatmap()` are generated and why they differ from the sorted data frames generated by `getMarkers()` would be appreciated. Thank you!

ArchR version = 1.0.2

**Comment by rcorces** [comment]
the `spmat` line that you are referring to is this:
https://github.com/GreenleafLab/ArchR/blob/f6c0388bd37023400794c9ae8562ad69e3ba9fd7/R/MarkerFeatures.R#L969

I believe that this is just a way to measure how "unique" a peak is to a given cell grouping. A row is a gene and by normalizing by the rowSums you get how many cell groups have this marker in `passMat`. The markers get ordered by this normalized value which doesnt really have any bearing on the log2FC etc.
